### DISCUSSION TITLE
Checking in on TS/JS cloud functions

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#1139] [Q&A]
Hey @ganigeorgiev I see that goja has made it into the core, does this mean you intend to pursue cloud functions after all?

I wasn't successful at getting that approach to work, so for PocketHost I have an experimental branch where users can upload Deno bundles and use the JS SDK. This is actually a pretty nice dx because then there is only one API to learn. I can also handle REST endpoints that way too.

Anyway just curious how far you've gotten. For a good sense of something I had a lot of trouble figuring out, try adding a REST endpoint and handler via JS.

**Comment by ganigeorgiev** [comment]
> does this mean you intend to pursue cloud functions after all?

Not exactly, or at least not what you probably mean by cloud functions.

Currently goja is used only for loading migrations that looks something like this:

```js
// pb_migrations/1669663597_updated_posts_old.js
migrate((db) => {
  // up
  const dao = new Dao(db)
  const collection = dao.findCollectionByNameOrId("lngf8rb3dqu86r3")
  collection.name = "posts_new"
  return dao.saveCollection(collection)
}, (db) => {
  // down
  const dao = new Dao(db)
  const collection = dao.findCollectionByNameOrId("lngf8rb3dqu86r3")
  collection.name = "posts_old"
  return dao.saveCollection(collection)
})
```

---

In the near future I have plans to allow also binding to the app event hooks (eg. registering new routes or doing data validations). But this may not be available with the next v0.9 release.

I played with goja and v8go and I didn't have too much difficulties with registering new routes in both, but documenting the API and all wrappers will definitely take time.

Initially the implementation will be similar to the Go API and will also require restarting the server on change. This is because:

- adding routes is not a concurrent safe operation (we can fix this by adding a global mutex, but it may complicate too much the internals especially if there is a running db operation)
- because of the above, once a route is registered, currently there is no "safe" way to remove it
- similar to the above, but it is a little bit easier, currently we don't support removing hook listeners (it will be added in the near future)

Eventually, if we manage to solve the above problems, we may use [fsnotify](https://github.com/fsnotify/fsnotify) or similar to listen for file changes and dynamically update the attached JS routes and hooks. And eventually if there is a users demand for this, we may simulate "cloud functions" behavior by allowing updating the js files with a web API. But this currently is out of the scope.
